src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -321,19 +326,98 @@ isAcyclic = isRight . topSort
+        , innerGraphs   :: IntMap.IntMap (List (AdjacencyMap a))
+        , innerEdges    :: [(Int,(a,a))]
+        , outerEdges    :: [(a,a)]
+        } deriving (Show)
+
+gabowSCC :: Ord a => AdjacencyMap a -> State (StateSCC a) ()
+gabowSCC g =
+  do let dfs u = do p_u <- enter u
+                    forEach (postSet u g) $ \v -> do
+                      preorderId v >>= \case
+                        Nothing  -> do
+                          updated <- dfs v
+                          if updated then outedge (u,v) else inedge (p_u,(u,v))
+                        Just p_v -> do
+                          scc_v <- hasComponent v
+                          if scc_v
+                            then outedge (u,v)
+                            else popBoundary p_v >> inedge (p_u,(u,v))
+                    exit u
+     forM_ (vertexList g) $ \v -> do
+       assigned <- hasPreorderId v
+       unless assigned $ void $ dfs v
   where
-    Typed.GraphKL g decode _ = Typed.fromAdjacencyMap m
-    sccs      = map toList (KL.scc g)
-    leader    = Map.fromList [ (decode y, x)      | x:xs <- sccs, y <- x:xs ]
-    component = Map.fromList [ (x, expand (x:xs)) | x:xs <- sccs ]
-    expand xs = fromJust $ NonEmpty.toNonEmpty $ induce (`Set.member` s) m
-      where
-        s = Set.fromList (map decode xs)
+    -- called when visiting vertex v. assigns preorder number to v,
+    -- adds the (id, v) pair to the boundary stack b, and adds v to
+    -- the path stack s.
+    enter v = do SCC pre scc bnd pth pres sccs gs es_i es_o <- get
+                 let pre' = pre+1
+                     bnd' = (pre,v):bnd
+                     pth' = v:pth
+                     pres' = Map.insert v pre pres
+                 put $! SCC pre' scc bnd' pth' pres' sccs gs es_i es_o
+                 return pre
+
+    -- called on back edges. pops the boundary stack while the top
+    -- vertex has a larger preorder number than p_v.
+    popBoundary p_v = modify'
+      (\(SCC pre scc bnd pth pres sccs gs es_i es_o) ->
+         SCC pre scc (dropWhile ((>p_v).fst) bnd) pth pres sccs gs es_i es_o)
+
+    -- called when exiting vertex v. if v is the bottom of a scc
+    -- boundary, we add a new SCC, otherwise v is part of a larger scc
+    -- being constructed and we continue.
+    exit v = do newComponent <- (v==).snd.head <$> gets boundaryStack
+                when newComponent $ insertComponent v
+                return newComponent
+
+    insertComponent v = modify'
+      (\(SCC pre scc bnd pth pres sccs gs es_i es_o) ->
+         let gs' = IntMap.insert scc g_i gs
+             sccs' = List.foldl' (\sccs x -> Map.insert x scc sccs) sccs curr
+             scc' = scc + 1
+             bnd' = tail bnd
+             p_v = fst $ head bnd
+             g_i = fromList (uncurry edge.snd <$> es) <> fromList (vertex <$> curr)
+             (es,es_i') = span ((>=p_v).fst) es_i
+             pth' = tail $ dropWhile (/=v) pth
+             curr = v:takeWhile(/=v) pth
```

Comment:
I think you could use `span` or `break` here instead of traversing the list twice with `dropWhile` and `takeWhile`.

Reply:
I would also appreciate if you reorder this `let` sequence in the order of execution, otherwise, it's difficult to read.

Reply:
The shameful reason I originally wrote with `takeWhile`+`dropWhile` was to avoid the incomplete pattern match. What's needed is something like `splitAfter (==v)` which would put `v` inside the first field of the tuple. That, or a way to express that `v` is in the list with types, which I'm not sure how